app/src/main/java/io/stormbird/wallet/service/KeyService.java

```diff
@@ -0,0 +1,1412 @@
+                getAuthenticationForSignature();
+                break;
+            case NOT_DEFINED:
+            case TEXT_MARKER:
+            case WATCH:
+                signCallback.GotAuthorisation(false);
+                break;
+        }
+    }
+
+    /**
+     * Upgrade key security
+     *
+     * Works much the same as the import functions, but care must be taken not to destroy the existing key during upgrade
+     * In the process of attempting to encrypt using an auth-locked key, we necessarily overwrite the existing key.
+     * By attempting encryption we may trigger UserNotAuthenticatedException. In this case, the old key is already overwritten.
+     * Therefore, the flow re-writes the old keystore password/mnemonic bytes using a non-authentication locked AES key.
+     *
+     * The flow asks for authentication, if provided it will restore the password/mnemonic bytes from the key generated stored in the last step
+     * and again attempt to store the password/mnemonic bytes using an authentication-locked key. This time it will pass and the key is created.
+     * If Authentication fails there is a usable fallback key available.
+     *
+     *
+     * @param key
+     * @param callingActivity
+     * @param callback
+     * @return
+     */
+    public UpgradeKeyResult upgradeKeySecurity(String key, Activity callingActivity, SignAuthenticationCallback callback)
+    {
+        signCallback = callback;
+        activity = callingActivity;
+        callback.setupAuthenticationCallback(this);
+        //first check we have ability to generate the key
+        if (!deviceIsLocked())
+            return UpgradeKeyResult.NO_SCREENLOCK;
+
+        currentKeyAddress = key;
+
+        return upgradeKey();
+    }
+
+    /**
+     * SignData
+     *
+     * Flow for this function is by necessity simpler - this function is called from code that doesn't have access to an Activity, so can't create
+     * any signing dialog. The authentication event must be generated prior to entering the signing flow.
+     *
+     * If HDWallet - decrypt mnemonic, regenerate private key, generate digest, sign digest using Trezor libs.
+     * If Keystore - fetch keystore JSON file, decrypt keystore password, regenerate Web3j Credentials and sign.
+     *
+     * @param key
+     * @param transactionBytes
+     * @return
+     */
+    public synchronized byte[] signData(String key, byte[] transactionBytes)
+    {
+        currentKeyAddress = key;
+        Wallet wallet = new Wallet(currentKeyAddress);
+        wallet.checkWalletType(context);
+        switch (wallet.type)
+        {
+            case KEYSTORE_LEGACY:
+                return signWithKeystore(wallet, transactionBytes);
+            case KEYSTORE:
+                return signWithKeystore(wallet, transactionBytes);
+            case HDKEY:
+                String mnemonic = unpackMnemonic(currentKeyAddress, Operation.SIGN_DATA);
+                if (mnemonic.length() == 0)
```

Comment:
If mnemonic is length 0, a 4 byte array is returned which is used in MarketQueueService.sigFromByteData. In particular position 64 in this array is accessed which should throw an ArrayIndexOutOfBoundsException. Is this properly handled somewhere?